rrmpg/models/cemaneigehystgr4j.py

```diff
@@ -0,0 +1,701 @@
+            
+    return loss_value
+   
+def _loss_Q_SCA(X, *args):
+    """Return the loss value for the current parameter set for calibration on discharge and SCA."""
+    # Unpack static arrays
+    obs = args[0]
+    prec = args[1]
+    mean_temp = args[2]
+    frac_solid_prec = args[3]
+    etp = args[4]
+    NDSI1 = args[5]
+    NDSI2 = args[6]
+    NDSI3 = args[7]
+    NDSI4 = args[8]
+    NDSI5 = args[9]
+    snow_pack_init = args[10]
+    thermal_state_init = args[11]
+    sca_init = args[12] 
+    s_init = args[13] 
+    r_init = args[14] 
+    dtype = args[15]
+    loss_metric = args[16]
+    
+    # Create custom numpy array of the model parameters
+    params = np.zeros(1, dtype=dtype)
+    params['CTG'] = X[0]
+    params['Kf'] = X[1]
+    params['Thacc'] = X[2]
+    params['Rsp'] = X[3]
+    params['x1'] = X[4]
+    params['x2'] = X[5]
+    params['x3'] = X[6]
+    params['x4'] = X[7]
+    
+    # Calcuate simulated outflow
+    outflow, _, _, _, _, sca, _ = run_cemaneigehystgr4j(prec, mean_temp, etp,
+                                            frac_solid_prec, snow_pack_init, 
+                                            thermal_state_init, sca_init, s_init, r_init,
+                                            params[0])
+    
+    
+    ## Extract and flatten simulated snow-covered area (SCA) for each elevation band
+    sca1, sca2, sca3, sca4, sca5 = (
+        sca[:, 0].flatten() * 100, 
+        sca[:, 1].flatten() * 100,  
+        sca[:, 2].flatten() * 100,
+        sca[:, 3].flatten() * 100,
+        sca[:, 4].flatten() * 100,
+    )
+
+    # Choose loss function based on loss_metric argument
+    if loss_metric == "mse":
+        loss_q = calc_mse1(obs, outflow)
+        loss_sca1 = calc_mse1(NDSI1, sca1)
+        loss_sca2 = calc_mse1(NDSI2, sca2)
+        loss_sca3 = calc_mse1(NDSI3, sca3)
+        loss_sca4 = calc_mse1(NDSI4, sca4)
+        loss_sca5 = calc_mse1(NDSI5, sca5)
+    elif loss_metric == "kge":
+        loss_q = 1 - kge1(obs, outflow) 
+        loss_sca1 = 1 - kge1(NDSI1, sca1)
+        loss_sca2 = 1 - kge1(NDSI2, sca2)
+        loss_sca3 = 1 - kge1(NDSI3, sca3)
+        loss_sca4 = 1 - kge1(NDSI4, sca4)
+        loss_sca5 = 1 - kge1(NDSI5, sca5)
+    else:
+        raise ValueError("Invalid loss_metric. Choose 'mse' or 'kge'.")
+
+    ## Weighting factors: 75% on discharge, 5% on each SCA band
+    loss_value = (
+        0.75 * loss_q +
+        0.05 * loss_sca1 +
+        0.05 * loss_sca2 +
+        0.05 * loss_sca3 +
+        0.05 * loss_sca4 +
+        0.05 * loss_sca5
+    )
+    
+    return loss_value
+
+def calc_mse1(evaluation, simulation):
+    df = pd.DataFrame({"obs": evaluation, "sim": simulation}).dropna()
+    return calc_mse(df["obs"].values, df["sim"].values)
+
+def kge1(evaluation, simulation):
+    df = pd.DataFrame({"obs": evaluation, "sim": simulation}).dropna()
+    return kge(df["obs"].values, df["sim"].values)
```

Comment:
What is the reason for these two functions? You convert numpy arrays to a pandas dataframe only then to extract the data as a numpy array with .values. Why not directly call calc_mse (and calc_kge) in the if-elif-else statement above?